specs/phase2/shard-transition.md

```diff
@@ -0,0 +1,235 @@
+                        proposer_pubkey: BLSPubkey) -> ByteList[MAX_SHARD_BLOCK_SIZE]:
+    # We will add something more substantive in phase 2
+    return ByteList[MAX_SHARD_BLOCK_SIZE]()  # empty byte list
+```
+
+```python
+def get_shard_transactions(shard: Shard,
+                           slot: Slot,
+                           shard_state: ShardState,
+                           previous_beacon_root: Root,
+                           proposer_pubkey: BLSPubkey) -> Sequence[EETransaction]:
+    # Up to the proposer to collect transactions as it likes (based on EE preference, abstracted fee payments, etc.)
+    return []  # Placeholder
+```
+
+```python
+class ExecutionEnvironment(Protocol):
+    def init_witness(self,
+                     shard: Shard,
+                     slot: Slot,
+                     shard_state: ShardState,
+                     previous_beacon_root: Root,
+                     proposer_pubkey: BLSPubkey) -> ByteList[MAX_EE_WITNESS_SIZE]:
+        ...
+    
+    def verify_access_list(self, access_list) -> bool:
+        ...
+
+    # Add witness data to the state. Do not overwrite any branches, items, etc. in the witness if the first EE call has priority.
+    def merge_witness(self, prev:  ByteList[MAX_EE_WITNESS_SIZE], addition: ByteList[MAX_TRANSACTION_WITNESS_SIZE]) -> ByteList[MAX_EE_WITNESS_SIZE]:
+        ...
+
+    # False if witness_data is invalid for given ee_state_root.
+    # (i.e. witness verification bubbles up an exception / error return)
+    def verify_witness(self, ee_state_root: Root, witness_data: ByteList[MAX_TRANSACTION_WITNESS_SIZE]) -> bool:
+        ...
+
+    # Prepare an EE runner that is loaded with the given EE state witness. 
+    def prepare_runner(self, witness_data: ByteList[MAX_TRANSACTION_WITNESS_SIZE]) -> EnvironmentRunner:
+        ...
+```
+
+```python
+class EnvironmentHost(Protocol):
+    def shard_transfer(self, value: Gwei, recipient: Shard) -> bool:  # False if insufficient funds 
+        ...
+    def ee_call(self, ee_call: EECall) -> bool:  # Within single shard
+        ...
+    def slot(self) -> Slot:  # Example host environment variable available during execution
+        ...
+    # TODO: other EE host functions
+```
+
+```python
+class EnvironmentRunner(Protocol):
+    def check_transaction(self, tx: EETransaction) -> uint256:  # Account abstraction payment check, output fee value
+        ...  # TODO: define how we hit the "pay" opcode, how the first part is deterministic and pays the fee, etc.
+
+    # Make a call. If failing, return false and rollback state to its contents from before this call. 
+    def make_call(self, call_data: ByteList[MAX_TRANSACTION_CALL_SIZE], host: EnvironmentHost) -> (bytes, bool, bool):  # out: call output, call pass/fail, access ok
+        ...
+    # Return the latest state root
+    def ee_state_root(self) -> Root:
+        ...
+```
+
+```python
+EXECUTION_ENVIRONMENTS = {
+    0: ETH_1_ENVIRONMENT,
```

Comment:
Is this meant to be for the [eth1 on shard 0](https://ethresear.ch/t/alternative-proposal-for-early-eth1-eth2-merge/6666) proposal? Or the former roadmap around an [Eth1 EE](https://ethresear.ch/t/the-eth1-eth2-transition/6265)?